### ISSUE TITLE
For keyphrase extraction, which loss to use when I have some positive examples and no negative examples

### ISSUE THREAD

**Issue by MoonRiver-l** [UKPLab/sentence-transformers#572]
Hello, thanks a lot for sharing your amazing work!
I want to fine-tune a pre-trained model for keyphrase extraction.
In my datasets, there are some golden keyphrases for each text which are the positive examples and no negative examples. 
Like this:
```
text 1, golden keyphrase 1_1, golden keyphrase 1_2,..., golden keyphrase 1_n
text 2, golden keyphrase 2_1, golden keyphrase 2_2,..., golden keyphrase 2_m
text 3, golden keyphrase 3_1, golden keyphrase 3_2,..., golden keyphrase 3_k
```
For such data, is there a loss that can be used?
Maybe ContrastiveLoss?

Thanks
Kawamura

**Comment by nreimers** [comment]
You can try the BatchAllTripletLoss:
https://www.sbert.net/docs/package_reference/losses.html#batchalltripletloss

(And the other variations like BatchHardTripletLoss etc.)

What can also be interesting is this loss:
https://www.sbert.net/docs/package_reference/losses.html#multiplenegativesrankingloss

**Comment by MoonRiver-l** [comment]
Thanks for your comments!
I will try to use BatchAllTripletLoss which may be useful.
And about MultipleNegativesRankingLoss, I'm a little confused. 

`This loss expects as input a batch consisting of sentence pairs (a_1, p_1), (a_2, p_2)…, (a_n, p_n) where we assume that (a_i, p_i) are a positive pair and (a_i, p_j) for i!=j a negative pair.`

It seems that this loss is for data with only one positive sample for each anchor. But when I have multiple positive samples, this loss will still have a good performance? I'm a little confused.
Thanks
Kawamura

**Comment by nreimers** [comment]
It depends how much data you have. You must ensure that there is only one positive instance per batch. What is the chance that there are multiple positives from the same example in the same batch? If the chance is small, you can go ahead and use it.

Otherwise you can also a data sampler, which constructs the batch with the needed property:
https://pytorch.org/docs/stable/data.html

**Comment by MoonRiver-l** [comment]
Thank you very much. I will check.

**Comment by DahaeKimEsther** [comment]
@nreimers 
Hello, thanks a lot to share your work so that I'm using it for our team project in a company.
However, it seems like I've been using Multiple Negatives Ranking Loss in a wrong way when applying it to mulitple positives
since the similarities of anchor and positives decreased after 10 epochs of training, e.g. 0.87 to 0.13
the result was the same when I made the batch with only one positive as mentioned above while using Dataset customed and changing collate function of SentenceTransformer.py 

We have data from an library which collect data of books from all over the city
the data consists of an anchor and multiple positives like below
`representitive title of book A, variant title1 of book A, ..., variant title N of book A`
there are about 5million books and book A is one of it as an example.

the problem is that there are some wrong variant title of different book.
`representitive title of book A, variant title1 of book A, similar title of book B,..., variant title N of book A`
this happend as a limitation of library's current system.

it was impossible to give thresholds to each bundle of book data 
to remove title of different bookB in book A in the example
because the number of variant titles are about 26million.

After that our team determined to train both type of data as anchor and multiple positives.
`representitive title of book A, variant title1 of book A, ..., variant title N of book A`
`representitive title of book A, variant title1 of book A, similar title of book B,..., variant title N of book A`

Our team expected that cosine similarity between representitive title of book A and variant title1 of book A would converge to above 0.9 while cosine similarity of representitive title of book A and similar title of book B is below 0.9 e.g. 0.7, 0.8.

Q1. Is it right to use MNR Loss to use this case?
Q2. Would it be better if we use the classification model while setting unique label for each book?
Q3. I've tried Batch All Triplet Loss with SentenceLabelDatset which was mentioned at [#636](https://github.com/UKPLab/sentence-transformers/issues/636#issuecomment-749132363) to include every 2 samples in each batch and same thing happened(the cosine similarity decreased after training). Should I avoid to use the Dataset I used?

Thanks for reading!
I'll do my best to explain if there is ambiguous english expressions
and I would reply myself if I solve the problem.
